### PR TITLE
fix(application): support snap/craft environment variables

### PR DESCRIPTION
**Pull request by mattculler**
- [x] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [x] Have you successfully run `tox run -m lint`?
- [x] Have you successfully run `tox run -e test-py310`? (supported versions: `py39`, `py310`, `py311`, `py312`)

-----

### CODE REVIEW

**Comment by mattculler** [comment]
Ran the unit tests and linters but forgot about the spread tests, will fix and add the commit to this PR unless someone tells me otherwise.

**Comment by tigarmo** [comment]
@mattculler here's a traceback from the spread failures to help you (the output has a lot of noise):

```
2024-02-22T21:11:50.5795366Z 2024-02-22 21:11:50 Error executing google:ubuntu-22.04-64:tests/spread/extensions/ros2-humble-meta-hello:colcon_humble_ros_base (feb222109-634932) : 
2024-02-22T21:11:50.5798064Z -----
2024-02-22T21:11:50.5799863Z + cd ../../plugins/craft-parts/build-and-run-hello/colcon-ros2-wrapper
2024-02-22T21:11:50.5801222Z + snapcraft
2024-02-22T21:11:50.5802308Z Traceback (most recent call last):
2024-02-22T21:11:50.5803539Z   File "/snap/snapcraft/x1/bin/snapcraft", line 8, in <module>
2024-02-22T21:11:50.5804744Z     sys.exit(main())
2024-02-22T21:11:50.5807229Z   File "/snap/snapcraft/x1/lib/python3.10/site-packages/snapcraft/application.py", line 252, in main
2024-02-22T21:11:50.5808899Z     app = Snapcraft(
2024-02-22T21:11:50.5810748Z   File "/snap/snapcraft/x1/lib/python3.10/site-packages/snapcraft/application.py", line 130, in __init__
2024-02-22T21:11:50.5812327Z     craft_cli.emit.debug(
2024-02-22T21:11:50.5813918Z   File "/snap/snapcraft/x1/lib/python3.10/site-packages/craft_cli/messages.py", line 414, in func
2024-02-22T21:11:50.5815309Z     raise RuntimeError("Emitter needs to be initiated first")
2024-02-22T21:11:50.5816385Z RuntimeError: Emitter needs to be initiated first
```

**Comment by tigarmo** [comment]
Can you rename the PR according to https://github.com/canonical/starbase/blob/main/HACKING.rst#commits ?

File: snapcraft/application.py [hunk 1]
```diff
@@ -121,6 +121,17 @@ def __init__(self, *args, **kwargs):
         # Whether we know that we should use the core24-based codepath.
         self._known_core24 = False
 
+        env_vars = {
+            ev: "SNAP" + ev for ev in ("CRAFT_BUILD_FOR", "CRAFT_BUILD_ENVIRONMENT")
+        }
+        for new_envvar, old_envvar in env_vars.items():
```
Comment (tigarmo) [inline snapcraft/application.py:127]:
```suggestion
        for craft_var, snapcraft_var in env_vars.items():
```

I think these names are more descriptive with regards to the source of the variables

File: tests/unit/test_application.py [hunk 2]
```diff
@@ -143,3 +144,21 @@ def test_build_planner_success_architecture_all(base, build_base, expected_base)
         assert build_info.platform == f"{expected_base.name}@{expected_base.version}"
 
     assert "all" not in [a.build_on for a in architectures]
+
+
+@pytest.mark.parametrize("new_envvar", {"CRAFT_BUILD_FOR", "CRAFT_BUILD_ENVIRONMENT"})
```
Comment (tigarmo) [inline tests/unit/test_application.py:149]: maybe we should put these variables in a top-level list in `snapcraft.application` and then use it both there and here

File: tests/unit/test_application.py [hunk 3]
```diff
@@ -143,3 +144,21 @@ def test_build_planner_success_architecture_all(base, build_base, expected_base)
...
+    old_envvar = "SNAP" + new_envvar
+    env_val = "woop"
+
+    monkeypatch.setenv(old_envvar, env_val)
+    assert os.getenv(new_envvar) is None
+
+    snapcraft_services = services.SnapcraftServiceFactory(app=application.APP_METADATA)
+    application.Snapcraft(app=application.APP_METADATA, services=snapcraft_services)
+
+    assert os.getenv(new_envvar) == env_val
```
Comment (tigarmo) [inline tests/unit/test_application.py:163]: thanks, looks good

**Review by tigarmo — CHANGES_REQUESTED** [review]

**Review by mr-cal — COMMENTED** [review]
I don't have any more feedback beyond what Tiago said, so just re-request me when it's ready for review again!

File: tests/unit/test_application.py (hunk 2 above)
Comment (mattculler) [reply tests/unit/test_application.py:149]: @tigarmo How about something like this in the top level of `snapcraft/application.py`?

```
MAPPED_ENV_VARS = {
    ev: "SNAP" + ev for ev in ("CRAFT_BUILD_FOR", "CRAFT_BUILD_ENVIRONMENT")
}   
```

**Review by lengau — APPROVED** [review]
Looks great, thanks!

**Review by tigarmo — APPROVED** [review]

**Comment by tigarmo** [comment]
@mattculler I merged mine first so now _you_ have to deal with the conflicts!! :sunglasses:

**Pull request merged**